enhancements/kube-apiserver/enforce-labels.md

```diff
@@ -0,0 +1,198 @@
+
+### Goals
+
+- enforce a per project (namespace) labeling scheme
+- ability to configure a cluster-level labeling scheme to enforce on all or a subset of namespaces
+- specify the labeling scheme on resource type and resource level
+- the labeling should complement installation time (eg. Helm or Operator) labeling scheme
+- ensure the labeling scheme is maintained over resource lifecycle (e.g. pod restart)
+- allow customer organizations to update the labeling scheme at any time
+- relabel existing resources upon labeling scheme update
+
+## Proposed solutions Gatekeeper
+
+Currently two solutions are being evaluated:
+
+- Gatekeeper
+- In-house solution
+
+### Gatekeeper
+
+https://github.com/open-policy-agent/gatekeeper/
+
+Gatekeeper is an opensource constraint policy enforcement tool.
+So far gatekeeper only provides a validation webhook, but has advanced plans for a mutation webhook. We're in discussion with gatekeeper people (smythe@google.com) about extending this to provide updates to existing resources in reaction to MutationTemplate changes.
+
+PROs:
+
+- advanced plans to provide a mutation webhook implementing a MutationTemplate (design doc [here|https://docs.google.com/document/d/1G7WgZKx1Y3VOTUjrqn7DjDaZgSKCIZowILm_I6psrw0/edit?pli=1#
+])
+- seem to be interested in adding a updates to existing resources in reaction to MutationTemplate changes (but no hard commitment on this)
+- a regularly release opensource project with  community backed by companies like google, ms, RH, ...
+
+CONs:
+- don't provide any functionality we need yet
+- reactive resource updates still uncertain
+- alpha version of the mutation webhook only available by the end of the year
+- an overkill for our immediate needs, provides a much broader functionality than we need
+
+### In-house solution
+
+If using gatekeeper is not feasible, a simple custom solution should be considered. Below is a proposal of such a project.
+
+PROs:
+
+- we own it, can add anything that is needed
+
+CONs:
+
+- still vaporware
+
+A mutating webhook (plus a  controller) will inspect created resources in specified namespaces for labeling scheme compliance. The webhook will check the namespace, type, ... of the created resource, match it to the configured labeling scheme, and add any missing labels/annotations. The webhook could also update incorrectly set labels/annotations if requested.
+
+The webhook will be configured by means of a LabelSchemeConfig CRD. The LabelSchemeConfig CRD will specify a target namespace, allowing to specify a label scheme for the specific namespace (an alternative solution of having the LabelSchemeConfig CRD namespaced would allow project admins to access the resource, which is not desirable).  
+ The controller will look for LabelSchemeConfig CRDs in it's own namespace.
+
+A sample LabelSchemeConfig resource is shown below:
+
+```yaml
+apiVersion: v1
```

Comment:
this needs a proper, fully qualified api group.

Reply:
Changed to: labelscheme.openshift.io/v1.
This is just a placeholder for now.